File: machine/ts/machine/runtime/vblank.ts
import { IRQ_VBLANK } from '../../spec/bmsx/io';
import {
	GX_GPU_PCRTC_RUNTIME_EDGE_VBLANK_BEGIN,
	GX_GPU_PCRTC_RUNTIME_EDGE_VBLANK_END,
} from '../devices/gx/gpu_pcrtc';
import type { FrameState } from './frame/state';
import { Runtime } from './runtime';

export class VblankState {
	private vblankSequence = 0;
	private lastCompletedVblankSequence = 0;
	private activeTickCompleted = false;

	constructor(private readonly runtime: Runtime) {
	}

	public get tickCompleted(): boolean {
		return this.activeTickCompleted;
	}

	public reset(): void {
		const runtime = this.runtime;
		this.vblankSequence = 0;
		this.lastCompletedVblankSequence = 0;
		this.activeTickCompleted = false;
		runtime.machine.inputController.cancelSampleArm();
		runtime.machine.irqController.postLoad();
	}

	public prepareRestore(): void {
		this.vblankSequence = 0;
		this.lastCompletedVblankSequence = 0;
		this.activeTickCompleted = false;
	}

	public beginTick(): void {
		this.activeTickCompleted = false;
	}

	public abandonTick(): void {
		this.activeTickCompleted = false;
	}

	public handleGpuRuntimeEdge(edge: number): void {
		if (edge === GX_GPU_PCRTC_RUNTIME_EDGE_VBLANK_BEGIN) {
			this.enterVblank();
			return;
		}
		if (edge === GX_GPU_PCRTC_RUNTIME_EDGE_VBLANK_END) {
			return;
		}
	}

	private enterVblank(): void {
		const runtime = this.runtime;
		this.vblankSequence += 1;
		runtime.machine.gxGpu.presentReadyFrameOnVblankEdge();
		runtime.machine.inputController.onVblankEdge(
			runtime.machine.scheduler.nowCycles,
		);
		runtime.machine.irqController.raise(IRQ_VBLANK);
		if (runtime.frameLoop.frameActive) {
			this.completeTickIfPending(runtime.frameLoop.frameState, this.vblankSequence);
		}
	}

	private completeTickIfPending(frameState: FrameState, vblankSequence: number): void {
		if (this.lastCompletedVblankSequence === vblankSequence) return;
		this.activeTickCompleted = true;
		const runtime = this.runtime;
		runtime.frameScheduler.enqueueTickCompletion(frameState);
		this.lastCompletedVblankSequence = vblankSequence;
	}
}

File: machine/cpp/machine/runtime/vblank.cpp
#include "machine/runtime/vblank.h"

#include "spec/bmsx/io.h"
#include "machine/devices/gx/gpu_pcrtc.h"
#include "machine/runtime/runtime.h"

namespace bmsx {

void VblankState::reset(Runtime& runtime) {
	m_vblankSequence = 0;
	m_lastCompletedVblankSequence = 0;
	m_activeTickCompleted = false;
	runtime.machine.inputController.cancelSampleArm();
	runtime.machine.irqController.postLoad();
}

void VblankState::prepareRestore() {
	m_vblankSequence = 0;
	m_lastCompletedVblankSequence = 0;
	m_activeTickCompleted = false;
}

void VblankState::beginTick() {
	m_activeTickCompleted = false;
}

void VblankState::abandonTick() {
	m_activeTickCompleted = false;
}

void VblankState::handleGpuRuntimeEdge(Runtime& runtime, u32 edge) {
	if (edge == GX_GPU_PCRTC_RUNTIME_EDGE_VBLANK_BEGIN) {
		enterVblank(runtime);
		return;
	}
	if (edge == GX_GPU_PCRTC_RUNTIME_EDGE_VBLANK_END) return;
}

void VblankState::enterVblank(Runtime& runtime) {
	m_vblankSequence += 1u;
	runtime.machine.gxGpu.presentReadyFrameOnVblankEdge();
	runtime.machine.inputController.onVblankEdge(
		runtime.machine.systemController.elapsedMilliseconds(),
		static_cast<u32>(runtime.machine.scheduler.nowCycles())
	);
	runtime.machine.irqController.raise(IRQ_VBLANK);
	if (runtime.frameLoop.frameActive) {
		completeTickIfPending(runtime, runtime.frameLoop.frameState, m_vblankSequence);
	}
}

void VblankState::completeTickIfPending(Runtime& runtime, FrameState& frameState, u64 vblankSequence) {
	if (m_lastCompletedVblankSequence == vblankSequence) return;
	m_activeTickCompleted = true;
	runtime.frameScheduler.enqueueTickCompletion(runtime, frameState);
	m_lastCompletedVblankSequence = vblankSequence;
}

} // namespace bmsx
